src/librustc_resolve/lib.rs

```diff
@@ -2697,20 +2699,50 @@ impl<'a> Resolver<'a> {
-                                }
+                    (Def::Struct(def_id), _) if ns == ValueNS && is_struct_like(def) => {
+                        if let Some((ctor_def, ctor_vis))
+                                = this.struct_constructors.get(&def_id).cloned() {
+                            let accessible_ctor = this.is_accessible(ctor_vis);
+                            if is_expected(ctor_def) && !accessible_ctor {
+                                err.span_label(span, format!("constructor is not visible \
+                                                              here due to private fields"));
+                            } else if accessible_ctor {
+                                let block = match ctor_def {
+                                    Def::StructCtor(_, CtorKind::Fn) |
+                                    Def::VariantCtor(_, CtorKind::Fn) => "(/* fields */)",
+                                    Def::StructCtor(_, CtorKind::Fictive) |
+                                    Def::VariantCtor(_, CtorKind::Fictive) => {
+                                        " { /* fields */ }"
+                                    }
+                                    def => bug!("found def `{:?}` when looking for a ctor",
+                                                def),
+                                };
+                                err.span_label(span, format!("did you mean `{}{}`?",
+                                                             path_str,
+                                                             block));
                             }
+                        } else {
+                            err.span_label(span, format!("did you mean `{} {{ /* fields */ }}`?",
+                                                         path_str));
                         }
-                        err.span_label(span, format!("did you mean `{} {{ /* fields */ }}`?",
-                                                     path_str));
                         return (err, candidates);
                     }
+                    (Def::VariantCtor(_, ctor_kind), _) if ns == ValueNS && is_struct_like(def) => {
+                        let block = match ctor_kind {
+                            CtorKind::Fn => "(/* fields */)",
+                            CtorKind::Fictive => " { /* fields */ }",
+                            def => bug!("found def `{:?}` when looking for a ctor",
+                                        def),
+                        };
+                        err.span_label(span, format!("did you mean `{}{}`?",
+                                                     path_str,
+                                                     block));
+                        return (err, candidates);
+                    }
+                    (Def::SelfTy(_, _), _) if ns == ValueNS && is_struct_like(def) => {
+                        err.note("can't instantiate `Self`, you must use the implemented struct \
```

Comment:
I don't think "instantiate" is a correct word here.
Maybe "can't use `Self` as a constructor"?

Reply:
Also the suggestion to use `Self { ... }` with braces is now missing (but maybe it's okay).

Reply:
Also suggestions to use `Foo { ... }` with braces are now missing for type aliases.